Add API for SELinuxRelabelPolicy

jsafrane opened the pull request:
SELinuxRelabelPolicy allows users to optionally skip recursive relabeling of all files on pod on volumes on systems with SELinux support.

KEP: https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1710-selinux-relabeling

When the feature is enabled + supported by volume plugin / CSI driver + enabled in `Pod.SecurityContext` + kubelet knows SELinux context of pod volumes, kubelet **mounts** pod's volumes with `-o context=<xyz>`. No relabeling is then necessary.

**What type of PR is this?**
/kind api-change

Changes from KEP:
`type SELinuxRelabelPolicy` -> `PodSELinuxRelabelPolicy`
Enum values `OnVolumeMount`, `AlwaysRelabel` -> `SELinuxRelabel*`

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Added SELinuxRelabelPolicy alpha feature gate, which allows pods to opt-in to skip recursive relabeling of pod's volumes when SELinux is enabled and mount the volume with the correct SELinux label directly.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]:  https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1710-selinux-relabeling
```

jsafrane committed:
Add API for skipping SELinux label change on pod startup.

jsafrane committed:
Add alpha feature gate SELinuxRelabelPolicy

jsafrane committed:
Generated API

jsafrane committed:
Add validation for SELinuxRelabelPolicy

jsafrane commented:
Preview of implementation of this feature is available in https://github.com/kubernetes/kubernetes/pull/91839 (All kubelet changes are there + AWS EBS and CSI volume plugin updates).

jsafrane commented:
cc @msau42 @tallclair @gnufied 
/label api-review

smarterclayton commented:
/assign @smarterclayton

tallclair commented on pkg/apis/core/types.go:
```diff
@@ -2891,6 +2907,16 @@ type PodSecurityContext struct {
+	// "OnVolumeMount" tries to mount volumes used by the Pod with the right context and skip recursive ownership
+	// change. Kubernetes may fall back to policy "Always" if a storage backed does not support this policy.
+	// + optional
+	SELinuxRelabelPolicy *PodSELinuxRelabelPolicy
```
I can't remember if this came up on the KEP, but should this be per-volume rather than per-pod?

jsafrane replied:
See below, I'm following FSGroupChangePolicy.

tallclair commented on pkg/apis/core/types.go:
```diff
@@ -2891,6 +2907,16 @@ type PodSecurityContext struct {
+	// "OnVolumeMount" tries to mount volumes used by the Pod with the right context and skip recursive ownership
+	// change. Kubernetes may fall back to policy "Always" if a storage backed does not support this policy.
+	// + optional
+	SELinuxRelabelPolicy *PodSELinuxRelabelPolicy
```
Can we group this & `FSGroupChangePolicy` together in a way that makes it clear both pertain to volumes? I'm thinking a sub-struct along the lines of `VolumeSecurityContext`? See above comment, it may make sense to move this to the volume struct.

jsafrane replied:
I just followed `fsGroupChangePolicy`, which, after the same discussion, ended up in SecurityContext. @gnufied, do you remember why we kept it there?

smarterclayton replied:
I would like to know why this wouldn't be in the volume struct too.

gnufied replied:
The discussion from `fsGroupChangePolicy` -  https://github.com/kubernetes/enhancements/pull/1478#discussion_r370904021 . tldr; of that discussion is - 

1. Having this setting on per-volume basis is more complicated for the end user and its puts this burden of deciding which policy to choose for which volume type. 
2. Also, if there were an automated tool which injected such policies in pods, the tool also has to be aware of volumes it should target.
3. On the other hand - having this setting on per-volume basis does not really buy much at the cost of added complexity.

tallclair commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -3153,6 +3153,22 @@ const (
+	// SELinuxRelabelOnVolumeMount indicates that, if supported by the storage backend, SELinux
+	// label will be assigned during mount of pod's volumes as a mount option. This can skip
+	// potentially long recursive SELinux label change performed by container runtime when starting
+	// a pod.
```
Why would you not always set this? Add the drawbacks to the doc comment.

jsafrane replied:
It's... complicated. I added more description there, however, I was not able to avoid mentioning Linux kernel.

smarterclayton replied:
What is the tradeoff when you skip this?  I.e. you talk about a benefit (performance) but you don't present what you lose (tim's question).  I would guess that you are not secure.  If you're not secure, why.  Is there a link somewhere to the presentation of why relabel is not desirable that we can copy?

jsafrane replied:
Sorry, I updated only storage/v1, forgetting core/v1.

Added "As tradeoff, all pods that simultaneously access the volumes must run with the same SELinux context to be able to access data on them, even if they use just a subPath of them."

The goal is to say it's safe, unless you use a single volume with two pods with different SELinux contexts, each using a different subpath of the volume. This was working before and it is working even now with policy "Always".

tallclair commented on pkg/apis/core/types.go:
```diff
@@ -2891,6 +2907,16 @@ type PodSecurityContext struct {
 	// sysctls (by the container runtime) might fail to launch.
 	// +optional
 	Sysctls []Sysctl
```
nit: this is in a different place in the versioned API. Can you swap it with FSGroupChangePolicy here?

jsafrane replied:
Reordered.

jsafrane committed:
Sync v1 and internal PodSecurityContext

Swap Sysctls and FSGroupChangePolicy in the internal API to have the
same order as in v1 API.

jsafrane committed:
Add examples to CSIDriver.SELinuxRelabelOnVolumeMount

jsafrane commented:
/close

`mount -o context=XYZ` is hard from a CSI driver container, see https://github.com/kubernetes/enhancements/pull/1879.
Reworking to merge with `fsGroupChangePolicy` in the linked PR.
